### ISSUE TITLE
font-family sanitization issue

### ISSUE THREAD

**Issue by jurajvalkucak** [OWASP/java-html-sanitizer#232]
Hi,

looks like there's issue with CSS font-family sanitization, when the input is first sanitized it adds quotes to font-families. When the sanitized content is sanitized again it removes some font-families and leaving blanks separated with commas, causing CSS font-family to be invalid.

**Input to sanitize:**
`<span style="font-family:WordVisi_MSFontService, Algerian, Algerian_EmbeddedFont, Algerian_MSFontService, sans-serif;">TEXT</span>`
	
**Sanitize input (adding quotes to font-families and lower case):**
`<span style="font-family:&#39;wordvisi_msfontservice&#39; , &#39;algerian&#39; , &#39;algerian_embeddedfont&#39; , &#39;algerian_msfontservice&#39; , sans-serif">TEXT</span>`
	
**Sanitize again (issue removing font-families and adding commas, causing invalid font-family tag):**
`<span style="font-family:, &#39;algerian&#39; , , , sans-serif">TEXT</span>`

The issue is caused if policy is configured like below:
`new HtmlPolicyBuilder().allowStyling(CssSchema.DEFAULT)`

Thanks,
Juraj

**Comment by mikesamuel** [comment]
I don't think we guarantee that sanitization is idempotent, but this looks like a bug.  The problem is probably somewhere in StylingPolicy and probably has to do with the underscores in the names that get removed.

**Comment by jtotht** [comment]
The condition at https://github.com/OWASP/java-html-sanitizer/blob/382c251b2aacae03fd2cfa76dffe01034f32b6bd/src/main/java/org/owasp/html/StylingPolicy.java#L158 means that if the font family name contains any character that is not alphanumeric, space, tab or hyphen (the latter since #90), nothing is emitted at all. This sounds wrong: if you encounter a quoted string, you should just emit it unchanged, at least as long as it doesn’t contain anything that has a special meaning ([which are `\ ' "` and newline](https://drafts.csswg.org/css-syntax-3/#string-token-diagram)). This code was introduced by you, @mikesamuel, in 24af1a69e2c9e4831a1c0f706d5bfdc1450b74fa, are there any chances you remembering why you chose this 13 years ago?

This problem is quite annoying, as Word apparently loves to include underscores in font names, so basically anything copied from Word (Word Online in our case) gets broken; due to the extra commas, it’s invalid CSS, so not even the fallback `sans-serif` is honored by browsers.

**Comment by jmanico** [comment]
Fixed in #431. Thanks for the report, and @jtotht for the diagnosis — it was exactly right.

The cause was an asymmetry between two paths. A font name was accepted **unquoted** but rejected once **quoted**, because the quoted path required every character to be alphanumeric, a space or a hyphen. Word emits names containing an underscore, so the first pass produced a name the second pass dropped, leaving the separator behind and, as you say, invalid CSS that costs you even the `sans-serif` fallback.

The quoted path now accepts what a font name legitimately contains — letters and digits in any script, spaces, and the hyphen, underscore and period. Your case round-trips:

```
font-family: WordVisi_MSFontService, Algerian, Algerian_EmbeddedFont, sans-serif
  -> font-family:'WordVisi_MSFontService' , 'Algerian' , 'Algerian_EmbeddedFont' , sans-serif
  -> (unchanged on a second pass)
```

To @mikesamuel's point above that idempotency was never guaranteed: it is now, at least for this property, and there is a fuzz check behind it. 60,000 adversarial `font-family` values built from quotes, backslashes, CSS escapes, comment delimiters, bidi overrides and CJK text, each sanitized twice — no input whose output changes on the second pass, and no output with unbalanced quotes.

Answering the question in your comment, @jtotht, about why the original condition was so strict: the backslash is the reason, and it is still rejected on purpose. The lexer normalizes a string's contents and hands the policy quotes and backslashes *already escaped* — `'it\27s'`, `'a\5c b'` — so re-emitting an escape verbatim would mean reasoning about escape parity to be sure the closing quote is still the closing quote. Rejecting it outright avoids that question. Everything else you would want in a font name is now allowed.

One extra thing came out of it: the unquoted path had no character test at all, so it would carry a format character such as U+202E RIGHT-TO-LEFT OVERRIDE into the output. Both paths now apply the same test.

**Comment by jmanico** [comment]
Correcting one thing I wrote above, because it overstates what changed.

I said the quoted path "now accepts what a font name legitimately contains — letters and digits in any script". That is true of the character test I changed, but misleading about the library, and a reader would reasonably conclude that Devanagari or Arabic font names now work. They do not.

What actually goes through, verified on `main`:

```
Latin, Cyrillic, Greek, Han, Hangul, kana        kept
Arabic / Hebrew / Devanagari / Thai base letters kept
  ...the same names carrying a combining mark    dropped
  (a Devanagari matra, an Arabic diacritic, Hebrew niqqud, a Thai vowel sign)
CJK extension B and other supplementary chars    dropped
```

Since most real Devanagari and Thai font names need those marks, those names still do not survive sanitization.

Two things worth being clear about:

- **This is not a limitation of the fix, and not new.** The CSS lexer drops those characters from a token before the policy ever runs. I checked against a build from before this fix and the results are identical, so nothing regressed here — it simply was not fixed, and relaxing the character test alone would not fix it.
- **Everything I claimed about your actual report still holds.** `WordVisi_MSFontService` and friends round-trip, sanitizing twice is stable, and the fuzz result stands.

#432 corrects the code comment to say this plainly and adds a test pinning both halves, so the claim fails if it stops being true.

If font names in a script that needs combining marks matter to you, that is a real gap and worth its own issue — say the word and I will open one rather than leaving it buried in this thread.
